utils/tinystr/src/int_ops.rs

```diff
@@ -98,6 +98,43 @@ impl Aligned4 {
         (invalid_case & 0x8080_8080) == 0x8080_8080
     }
 
+    pub const fn is_ascii_alphabetic_lowercase(&self) -> bool {
+        let word = self.0;
+        // See explanatory comments in is_ascii_alphabetic
+        let mask = (word + 0x7f7f_7f7f) & 0x8080_8080;
+        let lower = word | 0x2020_2020;
+        let alpha = !(lower + 0x1f1f_1f1f) | (lower + 0x0505_0505);
+        // See explanatory comments in is_ascii_lowercase
+        let invalid_case = !(word + 0x3f3f_3f3f) | (word + 0x2525_2525);
+        (alpha & mask) ^ (invalid_case & 0x8080_8080) == 0x8080_8080
```

Comment:
Notes:

- In `A = (alpha & mask)`, alpha characters are 0, and all others are 1
- In `B = (invalid_case & 0x8080_8080)`, uppercase alpha characters are 0, and all others are 1

Every byte ends up in one of the cells in the following table:

| | B=0 | B=1 |
|---|---|---|
| A=0 | alpha && upperalpha | alpha && !upperalpha |
| A=1 | !alpha && upperalpha | !alpha && !upperalpha |

The `^` operator selects cells `A=0/B=1` and `A=1/B=0`. Cell `A=0/B=1` is what we are looking for, and cell `A=1/B=0` has no elements, since `!alpha && upperalpha` is the empty set.

So, I believe what you wrote is correct.

However, see my other comment for a suggested optimization.

Reply:
Thank you for the diagram -- this is solid for my understanding on bit manipulation and I appreciate the table!